records/2022-04-07.reliability.service.configuration.md

```diff
@@ -0,0 +1,36 @@
+# Improve Notify service configuration reliability
+
+Date: 2022-04-07
+
+## Status
+
+**In review.**
+
+## Context
+
+The Notify API, Admin and Document Download API are configured using environment variables.  These environment variables are stored in encrypted `.env` files and Kubernetes manifests.  Overtime, misconfigurations have occurred because of inconsistencies in how the String values of the environment variables are converted to Boolean and Integer values.  
+
+The goal of this solution is to improve the reliability of service configuration by consistently setting configuration values across services and logging the service configuration when it starts so that it can be validated. 
+
+## Proposed solution
+
+There are two proposed solutions for this ADR.
+
+### Log configuration on service startup
+
+Update the Notify API, Admin and Document Download API to log their configuration values on startup.  This will allow us to validate that the running services have the expected configuration values.  Sensitive values will be redacted in the logs to prevent accidental exposure.
+
+### Use `environs` Python package
+
+Update the Notify API, Admin and Document Download API to use the [`environs` Python package](https://pypi.org/project/environs/) to load Boolean and Integer configuration values.  By using `environs` we will be able to take advantage of its `env.bool` and `env.int` functions to consistently load configuration values with default fallbacks.  Currently there are multiple different ways being used to coerce String configuration values into Booleans and Integers, which is leading to mis-configuation. 
+
+## Testing
+
+The logged output from the Notify services will be compared with the expected values in the encrypted `.env` files and Kubernetes manifests.
+
+## Additional considerations
+
+This proposal is related to the work being done in the following issues:
+- [cds-snc/notification-planning#421](https://github.com/cds-snc/notification-planning/issues/421)
+- [cds-snc/notification-planning#503](https://github.com/cds-snc/notification-planning/issues/503)
+- [cds-snc/notification-planning#504](https://github.com/cds-snc/notification-planning/issues/504)
```

Comment:
There is a template we used with previous ADRs that has an Outcome section, where we can come back a few months after on what we think this led to. I'd like to have one on this. It will be hard to measure but I think we will have a more solid configuration strategy, less mistakes and less incidents with these changes. What's your take on this?

Reply:
I'll get that added in - I think that's a good idea to have the `Outcome` section so we can actually check if our intent was achieved.

Reply:
Alright, I've added it!  I'll refine in a PR if you were thinking something different for this section.